velocity-engine-core/src/main/java/org/apache/velocity/util/introspection/MethodMap.java

```diff
@@ -387,9 +393,52 @@ public static Method getTopMostMethodDeclaration(Method method)
             }
             clazz = superClass;
         }
+
         return method;
     }
 
+    private static boolean canAccess(Method method)
+    {
+        // Check if the method is public
+        if (Modifier.isPublic(method.getModifiers())) {
+            if (method.isAccessible()) {
+                // The method accessible flag was already set to true so we assume we can call it
+                return true;
+            } else {
+                try {
+                    // Check if we are able to change the accessible flag
+                    if (trySetAccessible(method)) {
+                        // Restore the accessible flag to its former value
+                        method.setAccessible(false);
+
+                        // We were able to modify the accessible flag, so we should be able to call it
+                        return true;
+                    }
+                } catch (Exception e) {
+                    // We failed to change the accessible flag, so we won't be able to call the method
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean trySetAccessible(Method method)
+    {
+        boolean accessible = false;
+        if (trySetAccessible != null) {
+            try {
+                accessible = ((Boolean) trySetAccessible.invoke(method)).booleanValue();
+            } catch (Exception e) {
+                // Failed to call the trySetAccessible method, assume not accessible
+            }
+        } else {
+            method.setAccessible(true);
```

Comment:
So this call is only for Java 8, right? And it may throw, whereas the name of the method is `trySetAccessible()`, which would imply it does not throw. Why not move the try/catch from `canAccess()` here?

Reply:
Yes indeed, I remember thinking about it and then I totally forgot about it. Improving that.